include/ghex/unstructured/grid.hpp

```diff
@@ -19,15 +19,15 @@ namespace gridtools {
 
                 template<typename Index>
                 struct grid {
-                    using index_t = Index;
+                    using index_type = Index;
                 };
 
             } // namespace detail
 
             /** @brief type to indicate unstructured grids */
             struct grid {
                 template<typename Domain>
-                using type = detail::grid<typename Domain::index_t>;
+                using type = detail::grid<typename Domain::local_index_type>;
```

Comment:
```suggestion
                using index_type = typename Domain::local_index_type;
```
and instead of doing `grid::type::index_type` you just write `grid::index_type`. Am I missing something?

Reply:
Yep, my concern would only be on the typedef inside the high level `make_pattern` (`./include/ghex/pattern.hpp:122`). I can use `using grid_type = GridType`, but then I would need the domain type to infer `index_type`, and this is not known i.e. inside the specialized `pattern.hpp` class (see for ex `./include/ghex/unstructured/pattern.hpp:40`). I can use `index_type` to chose among the `make_pattern` specializations, but in this way I am no longer discriminating among structured / unstructured. I might be a bit biased, though :)

Reply:
we can either have the grid containing the domain and the domain the index
```
template <class Domain>
struct grid {
    using domain_type = Domain;
};
```
or we can have metafunctions to deduce the Domain:
```
template <typename T>
struct domain_of;

template <typename Domain>
struct domain_of< grid<Domain> > {
    using type = Domain;
};
```

Reply:
I guess the question is: id the index a property of the grid or of the domain?

Reply:
I think it should be a property of the Domain. My idea is: in the unstructured case by index type we mean the local index type. In the structured case the equivalent is the coordinate type. I think that in both cases this should be set on the application side (an thus defined in the domain descriptor) and GHEX should provide iteration spaces using the chosen coordinate / index type coherently.

Reply:
I think what I was not sure of was whether it would be fine to explicitly template the grid type on the domain type or not.

Reply:
Le my try to reason about it. A field is where we store data, and a field relates to the domain, I guess. A field has a storage, and this can be 1-2-3-4-5D whatever, even though the access function for the field entity should always be based on domain::index somehow. Either we say the partitioned grid defines the (partitioned) domain, or that the domain defines the grid, they seem pretty tight to one another to me. The grid has additional relational information that defines the neighborhood, while the domain is just the "space" where the fields live. The domain asks the grid for neighbors, I guess, so Maybe the domain should be derived from the Grid, and the fields from the domain...